### PR TITLE
build(all): use es2019 target to support on Node12

### PR DESCRIPTION
**Pull request by ijemmy**
## Description of your changes

See issue #899 on the bug description.

In target `es2020`, it support `?.` so the TS transpiler will not transform this syntax.

For example:

```typescript
//in target 2020
subsegment?.close();
```

will be:

```typescript
//in target 2019
subsegment === null || subsegment === void 0 ? void 0 : subsegment.close();
```

Previously, we have been building our packages on target `es2020`. When customers consume our modules, it will still have `?.`. If customers use NodeJS12 (which is `es2019`), it will throw an error `SyntaxError: Unexpected token '.'`. The `.` is the second character of `?.` syntax.

To fix this, we change all of the target to 2019. However, this cause a lot of breaks in our current code. 
1. Branch coverage will drop as `?.` is broken into a conditional statement. We need to add new test cases or tell TS that the value does exist by adding `!`. This is really case by case as the value cannot really be `undefined` in some case.
2. NodeJS12 doesn't have `crypto.randomUUID` that we use in e2e tests. So we have to add `uuid` as a `devDependency`

**Important note** Notice that one of the `pr-lint-and-test` fails.  This is intentional. I added the additional checks on NodeJS12 and NodeJS14. (We tested only on Node16). The failing one is Node12, and it fails on `examples/sam` package.  This package uses the npm published version of our library. And it really should fail as the currently published library is target 2020. Once we publish a new version, this should be fixed automatically.

### How to verify this change
Switch to NodeJS 12 and run all tests
```sh
nvm use lts/erbium #NodeJS 12 LTS
npm install # There will be some warnings about CDK only support NodeJS14+. But everything still works.
npm run lerna-test

# Switch AWS profile to run e2e test
npm run lerna-test:e2e
```

### Related issues, RFCs
#899 

### PR status

***Is this ready for review?:*** YES  
***Is it a breaking change?:*** NO

## Checklist

- [x] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [x] I have made corresponding changes to the *documentation*
- [x] I have made corresponding changes to the *examples*
- [x] My changes generate *no new warnings*
- [x] The *code coverage* hasn't decreased
- [x] I have *added tests* that prove my change is effective and works
- [ ] New and existing *unit tests pass* locally and in Github Actions
- [x] Any *dependent changes have been merged and published* in downstream module
- [x] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .github/workflows/pr_lint_and_test.yml [hunk 1]
```diff
@@ -7,12 +7,16 @@ jobs:
...
+      fail-fast: false
     steps:
       - uses: actions/checkout@v3
-      - name: "Use NodeJS 16"
+      - name: "Use NodeJS"
         uses: actions/setup-node@v3
         with:
-          node-version: '16'
+          node-version: ${{ matrix.version }}
       - name: Install npm@8.x
```
Comment (ijemmy) [inline .github/workflows/pr_lint_and_test.yml:20]: Some unit tests fail in NodeJS12, so I make use of matrix strategy here to make sure that we don't introduce any issue later.

File: packages/tracer/tests/unit/Tracer.test.ts [hunk 2]
```diff
@@ -8,11 +8,22 @@ import { Tracer } from '../../src';
...
 
+type CaptureAsyncFuncMock = jest.SpyInstance<unknown, [name: string, fcn: (subsegment?: Subsegment) => unknown, parent?: Segment | Subsegment]>;
+const createCaptureAsyncFuncMock = function(provider: ProviderServiceInterface): CaptureAsyncFuncMock {
+  return jest.spyOn(provider, 'captureAsyncFunc')
+    .mockImplementation((methodName, callBackFn) => {
+      const subsegment = new Subsegment(`### ${methodName}`);
+      jest.spyOn(subsegment, 'flush').mockImplementation(() => null);
+      callBackFn(subsegment);
+    });
+};
```
[the remark this answers is not in the document]
Comment (dreamorosi) [reply packages/tracer/tests/unit/Tracer.test.ts:25]: Thanks for the note and good spot.

**Comment by ijemmy** [comment]
e2e test passes: https://github.com/awslabs/aws-lambda-powertools-typescript/actions/runs/2416774907

**Review by dreamorosi — APPROVED** [review]
This is great!

Thanks for diving deep into this, I really really appreciate it.

**Review by flochaz — APPROVED** [review]
Great Job !

**Comment by ijemmy** [comment]
> node 12 tests failing :/

@flochaz It should fail until we release a new version that supports node12.

**Comment by flochaz** [comment]
> > node 12 tests failing :/
> 
> @flochaz It should fail until we release a new version that supports node12.

Ah yeah true :)

**Pull request merged**